### DISCUSSION TITLE
Closing a layer (modal) after form submit (with success)

### DISCUSSION THREAD

**Discussion by thiagomajesk** [unpoly/unpoly#510] [Q&A] [answered]
Hi! I've been reading some discussions here, but I'm still a little bit confused about how to achieve this behavior correctly.

The use case is fairly simple and I think it's mostly similar to what we have in the companies page for the Unpoly [demo](https://demo.unpoly.com/companies). However, I'm kinda frustrated that for some reason, it's not working as expected.

I have the following button that opens the "new.html" page into a modal:

```html
<a 
  href="/qK55VN2Wx4aGsfYTDxNfMm/admin/categories/new" 
  up-accept-location="/qK55VN2Wx4aGsfYTDxNfMm/admin/categories/$id/edit" 
  up-layer="new" 
  up-size="full" 
  up-target="#new-category-form">
  New Category
</a>
```

The "edit.html" page has a form that is rendered inside a section that looks like this:

```html
<section 
  id="new-category-form" 
  up-source="/qK55VN2Wx4aGsfYTDxNfMm/admin/categories/new" 
  up-time="false" 
  up-etag="false">
  <form 
    action="/qK55VN2Wx4aGsfYTDxNfMm/admin/categories" 
    method="post" 
    up-disable="" 
    up-submit="">
    <button type="submit">Save</button>
  </form>
</section>
```

As per usual, after the form submits to the designed action, if it fails, the page you are on is re-rendered and the form errors are displayed properly. If it succeeds the user is redirected to `/qK55VN2Wx4aGsfYTDxNfMm/admin/categories/:id/edit`.

However, it seems that the code I have in place is not working as I see in the CRUD example on the Unpoly website. What am I missing? I've tried setting different patterns for the `up-accept-location` URL, at first: `/*/admin/categories/:id/edit` and then: `/qK55VN2Wx4aGsfYTDxNfMm/admin/categories/$id/edit`.  Do I have to do anything on the server side as well?

After enabling Unpoly logging on my dev tools, I'm seeing the following error: 

```
up:form:submit Submitting form
unpoly.js:7399 up:request:load Loading POST /qK55VN2Wx4aGsfYTDxNfMm/admin/categories
unpoly.js:7399 up:request:loaded Loaded HTTP 200 response to POST /qK55VN2Wx4aGsfYTDxNfMm/admin/categories
unpoly.js:7399 up:fragment:loaded Loaded fragment from HTTP 200 response to POST /qK55VN2Wx4aGsfYTDxNfMm/admin/categories

Uncaught (in promise) up.CannotMatch: Could not find common target in current page and response (tried selectors [[up-main~=modal], [up-main~=overlay], [up-main=''], main, #new-category-form] in modal overlay)
```

The first thing that seems weird to me is that Unpoly seems to be trying to read a fragment from the POST response and not following the redirect to the edit action. In the demo, there's a lot more going on. What exactly am I doing wrong?

PS.: It seems that this is a very standard use case, should we have this better documented somewhere? Perhaps a recipes section with common use cases in the docs?

**Comment by adam12** [comment]
Firstly, can you confirm the version of Unpoly you're using? It helps orientate us if a local test is warranted.

Maybe I'm not following correctly, but if you're expecting a redirect, I wouldn't expect to see a `200` status code here. Can you confirm that your POST did indeed redirect? If it didn't, what did it return instead?

**Reply by thiagomajesk** [reply]
Hi, @adam12! Sorry, I completely forgot to include the version in the post... I'm using the latest `3.2.1`.

> Maybe I'm not following correctly, but if you're expecting a redirect, I wouldn't expect to see a 200 status code here. Can you confirm that your POST did indeed redirect? If it didn't, what did it return instead?

Yes, that's the weird part... I can confirm that this works if I load the page outside of the layer by going to the URL manually. The network request is the same when inside the layer:

![image](https://github.com/unpoly/unpoly/assets/13930952/aa1fd271-a73b-485f-b808-09bd75e691a7)

As you can see, it hits the controller and redirects to the edit page after the category is created. 

I see that you work with Elixir as well, so this is the controller code for reference (it's pretty basic):

```elixir
def create(conn, %{"category" => category_params}) do
    organization = conn.assigns.current_organization

    case Categories.create_category(organization, category_params) do
      {:ok, category} ->
        conn
        |> put_flash(:info, "Category create with success")
        |> redirect(to: ~p"/#{organization}/admin/categories/#{category}/edit")

      {:error, changeset} ->
        conn
        |> put_flash(:error, "Something wrong happened, check the errors bellow")
        |> render(:new, changeset: changeset)
    end
end
```

**Reply by adam12** [reply]
Everything looks OK from your screenshot and code. Perhaps this is the clue? `Uncaught (in promise) up.CannotMatch: Could not find common target in current page and response (tried selectors [[up-main~=modal], [up-main~=overlay], [up-main=''], main, #new-category-form] in modal overlay)
`

Is there anything else after that line in the logs? I'm not familiar with this section of code, but looking at the Demo app, Unpoly does use the 200 response but then (likely) discards it.

Do you have an element that matches that selector? It might be worth adding one if not.

**Reply by thiagomajesk** [reply]
This is the complete error message from the logs:

```
Uncaught (in promise) up.CannotMatch: Could not find common target in current page and response (tried selectors [[up-main~=modal], [up-main~=overlay], [up-main=''], main, #new-category-form] in modal overlay)
    at FromContent.cannotMatchTarget (http://localhost:4000/assets/app.js:13862:19)
    at FromContent.cannotMatchPostflightTarget (http://localhost:4000/assets/app.js:13845:18)
    at FromContent.execute (http://localhost:4000/assets/app.js:13797:71)
    at FromResponse.updateContentFromResponse (http://localhost:4000/assets/app.js:13661:72)
    at FromResponse.execute (http://localhost:4000/assets/app.js:13648:25)
    at FromURL.onRequestSettledWithResponse (http://localhost:4000/assets/app.js:13603:110)
    at FromURL.onRequestSettled (http://localhost:4000/assets/app.js:13597:27)
    at http://localhost:4000/assets/app.js:13573:69
    at async RenderJob.execute (http://localhost:4000/assets/app.js:12910:28)
```

> Do you have an element that matches that selector? It might be worth adding one if not.

If I add an `[up-target]` to the form (to be less specific about the fragment) like `up-target="#admin-content"` (wrapper to all page contents) or `up-target="html"` (the whole page) I still get the same error message. 

The edit action definitely returns a page that has those fragments, which makes me think if Unpoly is in fact following the redirect or trying to extract the fragment from the POST action.

**Reply by adam12** [reply]
Just to confirm, in the markup you render on all layouts (for sake of this example) have something similar to `[up-main]` to inform Unpoly what the main content element is?

I just want to eliminate that as the issue. 

https://unpoly.com/up-main

**Reply by thiagomajesk** [reply]
Hi, @adam12! I removed the `[up-target]` from the form and added `[up-main]` to the layout of the page which wraps the content, but still the same error message. I can confirm that the fragment is present in the return of the edit action.

![image](https://github.com/unpoly/unpoly/assets/13930952/58930ae7-881d-4f90-8f0b-b34833dbf14f)

**Comment by triskweline** [comment]
> If I add an [up-target] to the form (to be less specific about the fragment) like up-target="#admin-content" (wrapper to all page contents) or up-target="html" (the whole page) I still get the same error message.

Unpoly must match a selector in both the targeted layer and the response in order to update a fragment. Neither `#admin-content` nor `html` are within the targeted layer (the overlay).

> I removed the [up-target] from the form and added [up-main] to the layout of the page which wraps the content, but still the same error message. I can confirm that the fragment is present in the return of the edit action.

The `<main>` element from the screenshot is probably not contained by the overlay either.

> Unpoly seems to be trying to read a fragment from the POST response and not following the redirect to the edit action. 

Unpoly can only see the final response at the end of a redirect chain. This is a limitation of JavaScript.

**Reply by thiagomajesk** [reply]
Hi @triskweline! Let me see if I understood correctly... You are saying that Unpoly will try to replace fragments from the response into the modal itself and not the page that opened the layer - ok, the error now makes more sense to me.

So, If I understood it correctly, what I'm looking for is actually doing a full navigation after the redirect (that will also "close" the modal in the process). What do I have to do so Unpoly correctly follows the redirect (on success) instead of trying to replace fragments on the overlay?

> Unpoly can only see the final response at the end of a redirect chain. This is a limitation of JavaScript.

Yeah, that's what I was expecting, but the code was failing earlier because of the fragment (as you explained).

**Comment by thiagomajesk** [comment, marked as answer]
If anyone comes looking for an answer here, this is the proper solution to handle this use case:

- **Wiring things up**: Make sure that the `up-target` selector actually exists in all server responses (success and failures). It'll avoid that you spent countless hours trying to understand the pretty obvious error message telling you there's no common target (HTML node) between the current page and the response. As @triskweline explained in this thread previously, Unpoly expects that the target selector exists in the current **layer** and the response (the fragment replacement for layers works just like a page).

- **Check both ends are properly tied together**: After making sure fragments are being replaced correctly, you are going to notice that after submitting the form, the content from the redirected request (the edit page) is going to be placed inside the layer correctly. 

- **Closing the overlay**: You can instruct the overlay to be closed when your browser requests the page by specifying `up-accept-location` - I'm using: `*/admin/categories/$id/edit` to match a URL like: `5GQGZZQmxEoVMLM9WMmjm9/admin/categories/1/edit`. 

- **Rendering the correct page**: Now your overlay is closing properly but Unpoly is not showing the contents of the redirected page, it just shows the page you were before opening the layer. Fortunately, you can use `up-on-accepted` to control what is going to be rendered when the layer is closed (accepted): `up-on-accepted="up.render('#admin', { response: event.response })"` - again, make sure that the target exists in the response.

Now you have the following work-flow in place:

1. User clicks the "New Category" button
2. A modal opens with the contents from the new page
3. User submits the form (errors are shown in the layer itself)
  3.1 In case of errors, you should re-render the form and errors
  3.2 In case of success, you should redirect the user to the edit page
4. On success, the user is redirected to the edit page of the category

PS.: One thing that helped me understand a little bit more about how things work (besides the CRUD example) was this: https://github.com/unpoly/unpoly/discussions/498.

**Reply by adam12** [reply]
Thanks for coming back and clarifying. I'm still not sure what you did to actually solve this. Do you have a diff available?

I noticed that your target mentioned `#admin-content` yet the screenshot showed `#adminland-content`, which I wonder also contributed to the issue?

**Reply by thiagomajesk** [reply]
Hi, @adam12! I changed `#adminland-content` for `#admin-content` in the post, but the original code always had `#adminland-content` in it - sorry for the confusion.

The problem was exactly as the error said... The target needs to exist in the layer **and** the response. The problem was that the container for the form on the new page was `#new-category-form` and after the redirect (because we are now on the edit page), the container was `#edit-category-form`... I solved it by using just `#category-form` for the container target - which means it's now always present on the modal (new page) and the page after the redirect (edit page).

The code looks something like this (the parts that matter):

```html
<button
  up-layer="new"
  up-accept-location="*/adminland/categories/$id/edit"
  up-on-accepted="up.render('#adminland-content', { response: event.response, history: true })"
  up-target="#category-form"
  up-size="medium"
  up-animation="move-from-top"
>
  New Category
</button>
```

```html
<section id="category-form">
  <header><!-- [...]  --></header>
  <form up-target="#category-form" up-submit up-disable>
    <!-- [...]  -->
  </form>
</section>
```

PS.: This is working, but I still wish there was an easier way to instruct Unpoly to simply follow the redirect after having to manually render the contents of the page after the redirect.
